### DISCUSSION TITLE
nameserver-policy中返回rcode用于屏蔽广告，是否和fakeip模式冲突？

### DISCUSSION THREAD

**Discussion by reF1nd** [MetaCubeX/mihomo#668] [General]
看文档知道了`nameserver-policy`支持返回`rcode`，而`rcode`一般用于屏蔽广告域名的DNS查询。这里产生了一个疑问。
通过Wiki中[DNS解析流程页面](https://clash-meta.wiki/config/dns/diagram/)及FakeIP增强模式原理可知，当启用fakeip模式的情况下，应用们发起DNS查询时Clash会立即返回fakeip。

那么问题来了：`rcode`用于应用发起DNS查询时立即返回相应代码结束查询，用于屏蔽广告域名之类的DNS查询，但是在启用fakeip的情况下应用发起DNS查询会立即返回fakeip，只有在fakeip还原的域名在Clash内部进行DNS查询的时候才会匹配`nameserver-policy`中的规则，此时返回`rcode`相当于是对Clash内部返回的，而不是对应用返回的（因为此时应用已经得到fakeip了）。

所以`nameserver-policy`中返回`rcode`是否和fakeip模式冲突？还是说`rcode`相关规则优先级比fakeip还高？希望开发者能解答一下我的疑惑。

**Comment by yqs112358** [comment]
实测有效：把`nameserver-policy`里面返回`rcode`的 域名 / geosite / ruleset 在`fake-ip-filter`里面也写一遍

这样广告域名DNS解析时发现在`fake-ip-filter`中，行为会回退到`redir-host`，尝试进行本地DNS解析。又发现`nameserver-policy`中指定了返回`rcode`（`nameserver-policy`优先级高于`nameserver`/`fallback`），于是就会直接返回

**Reply by marlunzhang** [reply]
这样是不是会查询两遍广告列表

**Comment by PoseidonWake** [comment]
我也有疑问，一般广告规则都放在所有分流规则的最上边，一条广告域名发起连接，应该会直接匹配到规则reject，不需要dns解析，是不是到不了nameserver-policy这一流程？

**Reply by mengdegege** [reply]
是
